src/Avalonia.Controls/Flyouts/FlyoutBase.cs

```diff
@@ -0,0 +1,466 @@
+
+        private bool _isOpen;
+        private Control? _target;
+        protected Popup? _popup;
+        private FlyoutShowMode _showMode = FlyoutShowMode.Standard;
+        Rect? enlargedPopupRect;
+        IDisposable? transientDisposable;
+
+        /// <summary>
+        /// Gets whether this Flyout is currently Open
+        /// </summary>
+        public bool IsOpen
+        {
+            get => _isOpen;
+            private set => SetAndRaise(IsOpenProperty, ref _isOpen, value);
+        }
+
+        /// <summary>
+        /// Gets or sets the desired placement
+        /// </summary>
+        public FlyoutPlacementMode Placement
+        {
+            get => GetValue(PlacementProperty);
+            set => SetValue(PlacementProperty, value);
+        }
+
+        /// <summary>
+        /// Gets or sets the desired ShowMode
+        /// </summary>
+        public FlyoutShowMode ShowMode
+        {
+            get => _showMode;
+            set => SetAndRaise(ShowModeProperty, ref _showMode, value);
+        }
+
+        /// <summary>
+        /// Gets the Target used for showing the Flyout
+        /// </summary>
+        public Control? Target
+        {
+            get => _target;
+            private set => SetAndRaise(TargetProperty, ref _target, value);
+        }
+
+        public event EventHandler? Closed;
+        public event EventHandler<CancelEventArgs>? Closing;
+        public event EventHandler? Opened;
+        public event EventHandler? Opening;
+
+        public static FlyoutBase? GetAttachedFlyout(Control element)
+        {
+            return element.GetValue(AttachedFlyoutProperty);
+        }
+
+        public static void SetAttachedFlyout(Control element, FlyoutBase? value)
+        {
+            element.SetValue(AttachedFlyoutProperty, value);
+        }
+
+        public static void ShowAttachedFlyout(Control flyoutOwner)
+        {
+            var flyout = GetAttachedFlyout(flyoutOwner);
+            flyout?.ShowAt(flyoutOwner);
+        }
+
+        /// <summary>
+        /// Shows the Flyout at the given Control
+        /// </summary>
+        /// <param name="placementTarget">The control to show the Flyout at</param>
+        public void ShowAt(Control placementTarget)
+        {
+            ShowAtCore(placementTarget);
+        }
+
+        /// <summary>
+        /// Shows the Flyout for the given control at the current pointer location, as in a ContextFlyout
+        /// </summary>
+        /// <param name="placementTarget">The target control</param>
+        /// <param name="showAtPointer">True to show at pointer</param>
+        public void ShowAt(Control placementTarget, bool showAtPointer)
+        {
+            ShowAtCore(placementTarget, showAtPointer);
+        }
+
+        /// <summary>
+        /// Hides the Flyout
+        /// </summary>
+        /// <param name="canCancel">Whether or not this closing action can be cancelled</param>
+        public void Hide(bool canCancel = true)
```

Comment:
Should "canCancel" be part of the public API?
See: [https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.primitives.flyoutbase.hide?view=winrt-19041#Windows_UI_Xaml_Controls_Primitives_FlyoutBase_Hide](docs)

Reply:
Probably not. Will fix